File: app/src/main/java/jp/juggler/subwaytooter/table/MutedApp.java
package jp.juggler.subwaytooter.table;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.HashSet;

import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.util.LogCategory;

public class MutedApp {
	
	private static final LogCategory log = new LogCategory( "MutedApp" );
	
	public static final String table = "app_mute";
	public static final String COL_NAME = "name";
	private static final String COL_TIME_SAVE = "time_save";
	
	public static void onDBCreate( SQLiteDatabase db ){
		log.d( "onDBCreate!" );
		db.execSQL(
			"create table if not exists " + table
				+ "(_id INTEGER PRIMARY KEY"
				+ ",name text not null"
				+ ",time_save integer not null"
				+ ")"
		);
		db.execSQL(
			"create unique index if not exists " + table + "_name on " + table + "(name)"
		);
	}
	
	public static void onDBUpgrade( SQLiteDatabase db, int oldVersion, int newVersion ){
		if( oldVersion < 6 && newVersion >= 6 ){
			onDBCreate( db );
		}
	}
	
	public static void save( String app_name ){
		if( app_name == null ) return;
		try{
			long now = System.currentTimeMillis();
			
			ContentValues cv = new ContentValues();
			cv.put( COL_NAME, app_name );
			cv.put( COL_TIME_SAVE, now );
			App1.getDB().replace( table, null, cv );
			
		}catch( Throwable ex ){
			log.e( ex, "save failed." );
		}
	}
	
	public static Cursor createCursor(){
		return App1.getDB().query( table, null, null, null, null, null, COL_NAME + " asc" );
	}
	
	public static void delete( String name ){
		try{
			App1.getDB().delete( table, COL_NAME + "=?", new String[]{ name } );
		}catch( Throwable ex ){
			log.e( ex, "delete failed." );
		}
	}
	
	public static HashSet< String > getNameSet(){
		HashSet< String > dst = new HashSet<>();
		try{
			Cursor cursor = App1.getDB().query( table, null, null, null, null, null, null );
			if( cursor != null ){
				try{
					int idx_name = cursor.getColumnIndex( COL_NAME );
					while( cursor.moveToNext() ){
						String s = cursor.getString( idx_name );
						dst.add( s );
					}
				}finally{
					cursor.close();
				}
			}
		}catch( Throwable ex ){
			log.trace( ex );
		}
		return dst;
	}
	
	//	private static final String[] isMuted_projection = new String[]{COL_NAME};
	//	private static final String   isMuted_where = COL_NAME+"=?";
	//	private static final ThreadLocal<String[]> isMuted_where_arg = new ThreadLocal<String[]>() {
	//		@Override protected String[] initialValue() {
	//			return new String[1];
	//		}
	//	};
	//	public static boolean isMuted( String app_name ){
	//		if( app_name == null ) return false;
	//		try{
	//			String[] where_arg = isMuted_where_arg.get();
	//			where_arg[0] = app_name;
	//			Cursor cursor = App1.getDB().query( table, isMuted_projection,isMuted_where , where_arg, null, null, null );
	//			try{
	//				if( cursor.moveToFirst() ){
	//					return true;
	//				}
	//			}finally{
	//				cursor.close();
	//			}
	//		}catch( Throwable ex ){
	//			log.e( ex, "load failed." );
	//		}
	//		return false;
	//	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/table/MutedApp.kt
package jp.juggler.subwaytooter.table

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import jp.juggler.util.data.TableCompanion
import jp.juggler.util.data.getStringOrNull
import jp.juggler.util.data.queryAll
import jp.juggler.util.log.LogCategory

// リスト要素のデータ
class MutedApp(
    var id: Long = 0L,
    var name: String = "",
    var timeSave: Long = 0L,
) {
    companion object : TableCompanion {
        private val log = LogCategory("MutedApp")
        override val table = "app_mute"
        private const val COL_ID = "_id"
        private const val COL_NAME = "name"
        private const val COL_TIME_SAVE = "time_save"

        override fun onDBCreate(db: SQLiteDatabase) {
            log.d("onDBCreate!")
            db.execSQL(
                """create table if not exists $table
			($COL_ID INTEGER PRIMARY KEY
			,$COL_NAME text not null
			,$COL_TIME_SAVE integer not null
			)"""
            )
            db.execSQL(
                "create unique index if not exists ${table}_name on $table($COL_NAME)"
            )
        }

        override fun onDBUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            if (oldVersion < 6 && newVersion >= 6) {
                onDBCreate(db)
            }
        }
    }

    @Suppress("MemberVisibilityCanBePrivate")
    class ColIdx(cursor: Cursor) {
        val idxId = cursor.getColumnIndex(COL_ID)
        val idxName = cursor.getColumnIndex(COL_NAME)
        val idxTimeSave = cursor.getColumnIndex(COL_TIME_SAVE)

        fun readRow(cursor: Cursor) = MutedApp(
            id = cursor.getLong(idxId),
            name = cursor.getStringOrNull(idxName) ?: "",
            timeSave = cursor.getLong(idxTimeSave),
        )

        fun readOne(cursor: Cursor) = when {
            cursor.moveToNext() -> readRow(cursor)
            else -> null
        }

        fun readAll(cursor: Cursor) = buildList {
            while (cursor.moveToNext()) {
                add(readRow(cursor))
            }
        }
    }

    class Access(val db: SQLiteDatabase) {
        fun save(appName: String?) {
            if (appName == null) return
            try {
                val now = System.currentTimeMillis()
                val cv = ContentValues()
                cv.put(COL_NAME, appName)
                cv.put(COL_TIME_SAVE, now)
                db.replace(table, null, cv)
            } catch (ex: Throwable) {
                log.e(ex, "save failed.")
            }
        }

        fun delete(name: String) {
            try {
                db.delete(table, "$COL_NAME=?", arrayOf(name))
            } catch (ex: Throwable) {
                log.e(ex, "delete failed.")
            }
        }

        fun listAll() = db.queryAll(table, "$COL_NAME asc")
            ?.use { ColIdx(it).readAll(it) }
            ?: emptyList()

        fun nameSet() = buildSet {
            try {
                db.rawQuery("select $COL_NAME from $table", emptyArray())
                    ?.use { cursor ->
                        val idx_name = cursor.getColumnIndex(COL_NAME)
                        while (cursor.moveToNext()) {
                            add(cursor.getString(idx_name))
                        }
                    }
            } catch (ex: Throwable) {
                log.e(ex, "nameSet failed.")
            }
        }
        //	private static final String[] isMuted_projection = new String[]{COL_NAME};
        //	private static final String   isMuted_where = COL_NAME+"=?";
        //	private static final ThreadLocal<String[]> isMuted_where_arg = new ThreadLocal<String[]>() {
        //		@Override protected String[] initialValue() {
        //			return new String[1];
        //		}
        //	};
        //	public static boolean isMuted( String app_name ){
        //		if( app_name == null ) return false;
        //		try{
        //			String[] where_arg = isMuted_where_arg.get();
        //			where_arg[0] = app_name;
        //			Cursor cursor = App1.getDB().query( table, isMuted_projection,isMuted_where , where_arg, null, null, null );
        //			try{
        //				if( cursor.moveToFirst() ){
        //					return true;
        //				}
        //			}finally{
        //				cursor.close();
        //			}
        //		}catch( Throwable ex ){
        //			warning.e( ex, "load failed." );
        //		}
        //		return false;
        //	}
    }
}
